frontend/test/pytest/test_cuda_integration.py

```diff
@@ -502,5 +503,173 @@ def circuit2(a):
         assert_allclose(expected, observed)
 
 
+@pytest.mark.cuda
+class TestForLoops:
+    """Test for loops."""
+
+    def test_basic_loop(self):
+        """Test simple for loop."""
+
+        @qml.qnode(qml.device("lightning.qubit", wires=1))
+        def circuit_lightning(n):
+            @for_loop(0, n, 1)
+            def loop_fn(_):
+                qml.PauliX(0)
+
+            loop_fn()
+            return qml.state()
+
+        from catalyst.cuda import SoftwareQQPP
+
+        @qml.qnode(SoftwareQQPP(wires=1))
+        def circuit(n):
+            @for_loop(0, n, 1)
+            def loop_fn(_):
+                qml.PauliX(0)
+
+            loop_fn()
+            return qml.state()
+
+        cuda_compiled = catalyst.cuda.qjit(circuit)
+        catalyst_compiled = qjit(circuit_lightning)
+        expected = catalyst_compiled(2)
+        observed = cuda_compiled(2)
+        assert_allclose(expected, observed)
+
+    def test_loop_caried_values(self):
+        """Test for loop with updating loop carried values."""
+
+        @qml.qnode(qml.device("lightning.qubit", wires=1))
+        def circuit_lightning(n):
+            @for_loop(0, n, 1)
+            def loop_fn(_, x):
+                qml.PauliX(0)
+                return x + np.pi / 4
+
+            loop_fn(0.0)
+            return qml.expval(qml.PauliZ(0))
+
+        from catalyst.cuda import SoftwareQQPP
+
+        @qml.qnode(SoftwareQQPP(wires=1))
+        def circuit(n):
+            @for_loop(0, n, 1)
+            def loop_fn(_, x):
+                qml.PauliX(0)
+                return x + np.pi / 4
```

Comment:
This function doesn't use the value of `x`, so I don't think it could verify than x is updated correctly via its results.

But if follow @erick-xanadu's comment above then it might be unsupported anyways by CUDA and we should raise an error?

Reply:
Yes, I think it would be necessary to try out something with a loop carried dependency. If it works, then great, it means that the concrete evaluation of loops is working but the abstract evaluation needs more thought put into it.

Reply:
I've made the changes suggested by @erick-xanadu, I found that loop carried variable does not work at the moment. I think that makes sense since `cudaq` doesn't have a return statement support for loop carried variables in the loop body, so the loop carried variables do not get updated in the interpreter env.